src/app/shared/scroll/scroll-with-loading.directive.ts

```diff
@@ -0,0 +1,38 @@
+import {Directive, EventEmitter, HostListener, Output} from '@angular/core';
+import {LoadDirection} from '../load-direction/load-direction.enum';
+
+@Directive({
+    selector: '[appScrollWithLoading]',
+})
+export class ScrollWithLoadingDirective {
+    @Output() readonly loadData = new EventEmitter<LoadDirection>();
+
+    @HostListener('scroll', [
+        '$event.target.scrollHeight',
+        '$event.target.scrollTop',
+        '$event.target.offsetHeight',
+    ])
+    onScroll(scrollHeight: number, scrollTop: number, offsetHeight: number) {
+        const scroll = Math.round(scrollTop + offsetHeight);
+        const offset = scrollHeight - this.borderOffset;
+        const isScrollToTop = this.lastScrollTop > scrollTop;
+        const isNearTop = scrollTop < this.borderOffset;
+
+        if (isScrollToTop && isNearTop) {
+            this.loadData.emit(LoadDirection.ScrollTop);
+
+            return;
+        }
+
+        const isNearBottom = scroll > offset;
+
+        if (!isScrollToTop && isNearBottom) {
+            this.loadData.emit(LoadDirection.ScrollBottom);
+        }
+
+        this.lastScrollTop = scrollTop;
```

Comment:
Плюс данного подхода в том, что вычесления можно производить непосредсвенно перед блоком в котором они используются, по этому просто покажу, как это можно было бы сделать:
```suggestion
        const isScrollToTop = this.lastScrollTop > scrollTop;
        const isNearTop = scrollTop < this.borderOffset;

        if (isScrollToTop && isNearTop) {
            this.loadData.emit(LoadDirection.ScrollTop);

            return;
        }
        
        const scroll = Math.round(scrollTop + offsetHeight);
        const offset = scrollHeight - this.borderOffset;
        const isNearBottom = scroll > offset;

        if (!isScrollToTop && isNearBottom) {
            this.loadData.emit(LoadDirection.ScrollBottom);
        }

        this.lastScrollTop = scrollTop;
```